File: backend/app/schemas/diagnostics.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.common import Pagination


DiagnosticsStatus = Literal["healthy", "warning", "critical", "unknown"]
DiagnosticsSeverity = Literal["warning", "critical"]
DiagnosticsSource = Literal["server", "postgres", "backend", "frontend"]
DiagnosticsCommandEnvironment = Literal["local", "nginx", "docker"]
DiagnosticsEnvironmentMode = Literal["local", "docker", "nginx", "mixed", "unknown"]
DiagnosticsSourceKind = Literal["local_process", "docker_container", "proxy", "config", "unknown"]


class DiagnosticsCommandGroupOut(BaseModel):
    environment: DiagnosticsCommandEnvironment
    title: str
    commands: list[str] = Field(default_factory=list)
    warning: str | None = None


class DiagnosticsPortOut(BaseModel):
    port: int
    host: str
    state: str
    pid: int | None = None
    process_name: str | None = None
    command_line: str | None = None
    service: str | None = None
    detected_service: str | None = None
    port_role: str | None = None
    owner_role: str | None = None
    source_kind: DiagnosticsSourceKind = "local_process"
    is_primary_listener: bool = False
    explanation: str | None = None
    issues: list[str] = Field(default_factory=list)


class DiagnosticsProcessOut(BaseModel):
    pid: int
    parent_pid: int | None = None
    name: str
    service: str | None = None
    status: str | None = None
    command_line: str | None = None
    executable: str | None = None
    started_at: datetime | None = None
    uptime_seconds: int | None = None
    ports: list[int] = Field(default_factory=list)
    role: str | None = None
    source_kind: DiagnosticsSourceKind = "local_process"
    runtime_root_pid: int | None = None
    is_primary_runtime: bool = False
    is_auxiliary_runtime: bool = False
    explanation: str | None = None
    suspicious_reasons: list[str] = Field(default_factory=list)
    can_kill: bool = False


class DiagnosticsServiceOut(BaseModel):
    service: str
    display_name: str
    status: DiagnosticsStatus
    port: int | None = None
    host: str | None = None
    listener_pid: int | None = None
    http_ok: bool | None = None
    process_count: int = 0
    process_count_total: int = 0
    process_count_primary: int = 0
    process_count_auxiliary: int = 0
    warning_count: int = 0
    error_count: int = 0
    issues: list[str] = Field(default_factory=list)
    checked_at: datetime


class DiagnosticsDatabaseTableOut(BaseModel):
    table_name: str
    row_count: int = 0
    table_bytes: int = 0
    index_bytes: int = 0
    total_bytes: int = 0


class DiagnosticsDatabaseOverviewOut(BaseModel):
    database_name: str
    host: str
    port: int
    user: str
    database_bytes: int = 0
    table_count: int = 0
    total_rows: int = 0
    tables: list[DiagnosticsDatabaseTableOut] = Field(default_factory=list)
    issues: list[str] = Field(default_factory=list)


class DiagnosticsRuntimeNodeOut(BaseModel):
    key: str
    label: str
    source_kind: DiagnosticsSourceKind = "unknown"
    endpoint: str | None = None
    target: str | None = None
    status: DiagnosticsStatus = "unknown"
    details: list[str] = Field(default_factory=list)


class DiagnosticsRuntimeTopologyOut(BaseModel):
    environment_mode: DiagnosticsEnvironmentMode = "unknown"
    public_entrypoint: str | None = None
    frontend_url: str
    frontend_api_base: str
    backend_base_url: str
    backend_http_url: str | None = None
    backend_listener_pid: int | None = None
    backend_listener_port: int | None = None
    backend_http_ok: bool | None = None
    database_dsn: str
    database_host: str
    database_port: int
    database_name: str
    database_user: str
    is_frontend_backend_match: bool = True
    is_backend_database_local: bool = True
    status: DiagnosticsStatus = "unknown"
    nodes: list[DiagnosticsRuntimeNodeOut] = Field(default_factory=list)
    issues: list[str] = Field(default_factory=list)


class DiagnosticsSummaryOut(BaseModel):
    app_version: str
    checked_at: datetime
    host: str
    refresh_seconds: int = 3600
    environment_mode: DiagnosticsEnvironmentMode = "unknown"
    public_entrypoint: str | None = None
    services: list[DiagnosticsServiceOut] = Field(default_factory=list)
    ports: list[DiagnosticsPortOut] = Field(default_factory=list)
    database_overview: DiagnosticsDatabaseOverviewOut
    runtime_topology: DiagnosticsRuntimeTopologyOut
    process_count: int = 0
    warning_count: int = 0
    error_count: int = 0


class DiagnosticsLogEntryOut(BaseModel):
    id: str
    entry_id: str
    source: DiagnosticsSource
    severity: DiagnosticsSeverity
    signature: str
    category: str
    summary: str
    normalized_message: str | None = None
    raw_message: str
    observed_at: datetime | None = None
    file_path: str | None = None
    line_number: int | None = None
    is_low_signal: bool = False
    can_delete: bool = True
    possible_causes: list[str] = Field(default_factory=list)
    suggested_actions: list[str] = Field(default_factory=list)
    suggested_commands: list[DiagnosticsCommandGroupOut] = Field(default_factory=list)


class DiagnosticsLogsPageOut(Pagination[DiagnosticsLogEntryOut]):
    pass


class DiagnosticsDeleteLogsIn(BaseModel):
    entry_ids: list[str] = Field(default_factory=list, min_length=1)


class DiagnosticsDeleteLogsOut(BaseModel):
    deleted_count: int = 0
    missing_entry_ids: list[str] = Field(default_factory=list)


class DiagnosticsProcessKillOut(BaseModel):
    pid: int
    killed: bool
    message: str

File: frontend/src/api/diagnostics.ts
import { apiFetch } from "./client";
import { buildQuery, type Pagination } from "./entities";

export type DiagnosticsStatus = "healthy" | "warning" | "critical" | "unknown";
export type DiagnosticsSeverity = "warning" | "critical";
export type DiagnosticsSource = "server" | "postgres" | "backend" | "frontend";
export type DiagnosticsEnvironment = "local" | "nginx" | "docker";
export type DiagnosticsEnvironmentMode = "local" | "docker" | "nginx" | "mixed" | "unknown";
export type DiagnosticsSourceKind = "local_process" | "docker_container" | "proxy" | "config" | "unknown";

export type DiagnosticsCommandGroup = {
  environment: DiagnosticsEnvironment;
  title: string;
  commands: string[];
  warning: string | null;
};

export type DiagnosticsPort = {
  port: number;
  host: string;
  state: string;
  pid: number | null;
  process_name: string | null;
  command_line: string | null;
  service: string | null;
  detected_service: string | null;
  port_role: string | null;
  owner_role: string | null;
  source_kind: DiagnosticsSourceKind;
  is_primary_listener: boolean;
  explanation: string | null;
  issues: string[];
};

export type DiagnosticsProcess = {
  pid: number;
  parent_pid: number | null;
  name: string;
  service: string | null;
  status: string | null;
  command_line: string | null;
  executable: string | null;
  started_at: string | null;
  uptime_seconds: number | null;
  ports: number[];
  role: string | null;
  source_kind: DiagnosticsSourceKind;
  runtime_root_pid: number | null;
  is_primary_runtime: boolean;
  is_auxiliary_runtime: boolean;
  explanation: string | null;
  suspicious_reasons: string[];
  can_kill: boolean;
};

export type DiagnosticsService = {
  service: string;
  display_name: string;
  status: DiagnosticsStatus;
  port: number | null;
  host: string | null;
  listener_pid: number | null;
  http_ok: boolean | null;
  process_count: number;
  process_count_total: number;
  process_count_primary: number;
  process_count_auxiliary: number;
  warning_count: number;
  error_count: number;
  issues: string[];
  checked_at: string;
};

export type DiagnosticsDatabaseTable = {
  table_name: string;
  row_count: number;
  table_bytes: number;
  index_bytes: number;
  total_bytes: number;
};

export type DiagnosticsDatabaseOverview = {
  database_name: string;
  host: string;
  port: number;
  user: string;
  database_bytes: number;
  table_count: number;
  total_rows: number;
  tables: DiagnosticsDatabaseTable[];
  issues: string[];
};

export type DiagnosticsRuntimeTopology = {
  environment_mode: DiagnosticsEnvironmentMode;
  public_entrypoint: string | null;
  frontend_url: string;
  frontend_api_base: string;
  backend_base_url: string;
  backend_http_url: string | null;
  backend_listener_pid: number | null;
  backend_listener_port: number | null;
  backend_http_ok: boolean | null;
  database_dsn: string;
  database_host: string;
  database_port: number;
  database_name: string;
  database_user: string;
  is_frontend_backend_match: boolean;
  is_backend_database_local: boolean;
  status: DiagnosticsStatus;
  nodes: DiagnosticsRuntimeNode[];
  issues: string[];
};

export type DiagnosticsRuntimeNode = {
  key: string;
  label: string;
  source_kind: DiagnosticsSourceKind;
  endpoint: string | null;
  target?: string | null;
  status: DiagnosticsStatus;
  details: string[];
};

export type DiagnosticsSummary = {
  app_version: string;
  checked_at: string;
  host: string;
  refresh_seconds: number;
  environment_mode: DiagnosticsEnvironmentMode;
  public_entrypoint: string | null;
  services: DiagnosticsService[];
  ports: DiagnosticsPort[];
  database_overview: DiagnosticsDatabaseOverview;
  runtime_topology: DiagnosticsRuntimeTopology;
  process_count: number;
  warning_count: number;
  error_count: number;
};

export type DiagnosticsLogEntry = {
  id: string;
  entry_id: string;
  source: DiagnosticsSource;
  severity: DiagnosticsSeverity;
  signature: string;
  category: string;
  summary: string;
  normalized_message: string | null;
  raw_message: string;
  observed_at: string | null;
  file_path: string | null;
  line_number: number | null;
  is_low_signal: boolean;
  can_delete: boolean;
  possible_causes: string[];
  suggested_actions: string[];
  suggested_commands: DiagnosticsCommandGroup[];
};

export type DiagnosticsLogsPage = Pagination<DiagnosticsLogEntry>;

export type DiagnosticsDeleteLogsResult = {
  deleted_count: number;
  missing_entry_ids: string[];
};

export type DiagnosticsKillProcessResult = {
  pid: number;
  killed: boolean;
  message: string;
};

export async function getDiagnosticsSummary() {
  return apiFetch<DiagnosticsSummary>("/admin/diagnostics/summary");
}

export async function getDiagnosticsProcesses() {
  return apiFetch<DiagnosticsProcess[]>("/admin/diagnostics/processes");
}

export async function getDiagnosticsLogs(params: {
  source?: DiagnosticsSource | "";
  severity?: DiagnosticsSeverity | "";
  q?: string;
  include_low_signal?: boolean;
  date_from?: string;
  date_to?: string;
  page?: number;
  page_size?: number;
}) {
  const qs = buildQuery({
    source: params.source || undefined,
    severity: params.severity || undefined,
    q: params.q || undefined,
    include_low_signal: params.include_low_signal ? "true" : undefined,
    date_from: params.date_from || undefined,
    date_to: params.date_to || undefined,
    page: params.page ?? 1,
    page_size: params.page_size ?? 20
  });
  return apiFetch<DiagnosticsLogsPage>(`/admin/diagnostics/logs${qs}`);
}

export async function deleteDiagnosticsLogs(entryIds: string[]) {
  return apiFetch<DiagnosticsDeleteLogsResult>("/admin/diagnostics/logs/delete", {
    method: "POST",
    body: JSON.stringify({ entry_ids: entryIds })
  });
}

export async function killDiagnosticsProcess(pid: number) {
  return apiFetch<DiagnosticsKillProcessResult>(`/admin/diagnostics/processes/${pid}/kill`, {
    method: "POST"
  });
}
